File: app/src/main/java/mods/view/SimpleSpinner.java
package mods.view;

import static mods.constants.PreferenceKeys.DEFAULT_TRANSLATE_FROM_KEY;

import android.content.Context;
import android.graphics.drawable.ColorDrawable;
import mods.utils.LogUtils;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.Spinner;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import mods.constants.Constants;
import mods.preference.Prefs;
import mods.utils.StringUtils;

public class SimpleSpinner {

    private static final String TAG = "SimpleSpinner";

    private final Spinner spinner;
    private String selectedKey;

    public interface ISelectionCallback {
        void onItemSelected(String key);
    }

    private SimpleSpinner(Spinner spinner, String defaultChoice) {
        this.spinner = spinner;
        this.selectedKey = defaultChoice;
    }

    public static SimpleSpinner createSpinner(final Context context, SpinnerMap choices, String defaultChoice) {
        return createSpinner(context, choices, defaultChoice, null);
    }

    public static SimpleSpinner createSpinner(
            final Context context,
            SpinnerMap choices,
            String defaultChoice,
            @Nullable ISelectionCallback callback
    ) {
        Spinner spinner = new Spinner(context);
        spinner.setLayoutParams(new ViewGroup.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        );
        spinner.setTop(20);
        spinner.setBottom(20);
        spinner.setPadding(10, 10, 10, 10);
        spinner.setPopupBackgroundDrawable(new ColorDrawable(Colors.getThemeBackground()));

        List<String> keys = new ArrayList<>(choices.keySet());

        ArrayAdapter<String> adapter = new ArrayAdapter<>(context, Constants.SPINNER_LAYOUT, keys);
        adapter.setDropDownViewResource(Constants.SPINNER_DROP_DOWN_LAYOUT);

        spinner.setAdapter(adapter);

        SimpleSpinner simpleSpinner = new SimpleSpinner(spinner, defaultChoice);

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                String key = (String) parent.getItemAtPosition(position);
                simpleSpinner.setSelectedKey(key);
                LogUtils.log(TAG, "newValue = " + key);

                if (callback != null) {
                    callback.onItemSelected(key);
                } else {
                    Prefs.setString(DEFAULT_TRANSLATE_FROM_KEY, key);
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                // do nothing?
            }
        });

        LogUtils.log(TAG, "selection index = " + keys.indexOf(defaultChoice));

        spinner.setSelection(
                StringUtils.isEmpty(defaultChoice) || !keys.contains(defaultChoice)
                        ? 0
                        : keys.indexOf(defaultChoice)
        );

        return simpleSpinner;
    }

    public Spinner getSpinner() { return this.spinner; }

    public String getSelectedKey() { return this.selectedKey; }
    public void setSelectedKey(String key) { this.selectedKey = key; }

    @NonNull
    @Override
    public String toString() {
        return "SimpleSpinner{" +
                "spinner=" + spinner +
                ", selectedKey='" + selectedKey + '\'' +
                '}';
    }
}

File: app/src/main/java/mods/dialog/spinner/SimpleSpinner.kt
package mods.dialog.spinner

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.LinearLayout
import android.widget.Spinner
import mods.utils.LogUtils
import java.util.concurrent.atomic.AtomicBoolean
import androidx.core.graphics.drawable.toDrawable
import com.bluecord.R
import mods.view.Colors
import mods.view.dp

class SimpleSpinner private constructor(
    val spinner: Spinner,
    var selectedKey: String
) {
    private val firstSelection = AtomicBoolean(false)

    fun interface ISelectionCallback {
        fun onItemSelected(key: String, value: String)
    }

    override fun toString(): String {
        return "SimpleSpinner(spinner=$spinner, selectedKey=$selectedKey)"
    }

    companion object {
        private val TAG = SimpleSpinner::class.java.simpleName

        @JvmStatic
        fun create(
            context: Context,
            choices: SpinnerMap,
            defaultChoice: String,
            callback: ISelectionCallback
        ): SimpleSpinner {
            val keys = choices.keys.toList()

            val spinner = Spinner(context).apply {
                layoutParams = ViewGroup.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
                top = 10.dp()
                bottom = 10.dp()
                setPadding(6.dp(), 6.dp(), 6.dp(), 6.dp())
                setPopupBackgroundDrawable(Colors.getThemeBackground().toDrawable())
                adapter = ArrayAdapter(context, R.layout.blue_spinner, keys).apply {
                    setDropDownViewResource(R.layout.blue_spinner_dropdown)
                }
            }

            val simpleSpinner = SimpleSpinner(spinner, defaultChoice)

            spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                    // Ignore first selection, which is
                    // automatically done by setSelection() below
                    if (simpleSpinner.firstSelection.compareAndSet(false, true)) {
                        return
                    }
                    val key = parent.getItemAtPosition(position) as String
                    val value = choices[key]
                    simpleSpinner.selectedKey = key
                    LogUtils.log(TAG, "selectedKey=$key, selectedValue=$value")
                    callback.onItemSelected(key, value!!)
                }

                override fun onNothingSelected(parent: AdapterView<*>?) = Unit
            }

            // Call setSelection before setOnItemSelectedListener
            // This prevents the listener firing when first loaded
            val selectionIndex = if (defaultChoice.isEmpty() || !keys.contains(defaultChoice)) 0 else keys.indexOf(defaultChoice)
            spinner.setSelection(selectionIndex)
            return simpleSpinner
        }
    }
}
